File: Compiler/src/compiler/operators/GetClass.java
package compiler.operators;

import compiler.Operator;
import compiler.values.ClassValue;
import dictionaries.ClassesDictionary;
import util.*;

import java.util.ArrayList;
import java.util.List;

/**
 * Получить класс объекта
 * FIXME: А этот оператор нужен вообще?
 */
public class GetClass extends BaseOperator {

    /**
     * Имя предиката, используемое при компиляции
     */
    private static final String CLASS_PRED_NAME = "type";

    /**
     * Имя предиката, используемое при компиляции
     */
    private static final String SUBCLASS_PRED_NAME = "subClassOf";

    /**
     * Конструктор
     * @param args Аргументы
     */
    public GetClass(List<Operator> args) {
        super(args);
    }

    @Override
    public List<List<DataType>> argsDataTypes() {
        List<List<DataType>> result = new ArrayList<>();
        result.add(List.of(DataType.OBJECT));
        return result;
    }

    @Override
    public DataType resultDataType() {
        return DataType.CLASS;
    }

    @Override
    public CompilationResult compile() {
        // Объявляем переменные
        String value = "";
        String rulePart = "";
        String completedRules = "";

        // Получаем аргументы
        Operator arg0 = arg(0);

        // Компилируем аргументы
        CompilationResult compiledArg0 = arg0.compile();

        value = NamingManager.genVarName();

        rulePart = compiledArg0.rulePart();

        rulePart += JenaUtil.genTriple(
                compiledArg0.value(),
                JenaUtil.genRuleLink(JenaUtil.RDF_PREF, CLASS_PRED_NAME),
                value) +
                "noValue(" + value + "," + JenaUtil.genRuleLink(JenaUtil.RDF_PREF, SUBCLASS_PRED_NAME) + ")";

        completedRules = compiledArg0.completedRules();

        usedObjects = List.of(compiledArg0.value());

        return new CompilationResult(value, rulePart, completedRules);
    }
}

File: Compiler/src/compiler/operators/GetClass.kt
package compiler.operators

import compiler.Operator
import compiler.util.CompilationResult
import util.DataType
import util.JenaUtil
import util.JenaUtil.genLink
import util.JenaUtil.genNoValuePrim
import util.JenaUtil.genTriple
import util.NamingManager.genVarName

/**
 * Получить класс объекта
 */
class GetClass(args: List<Operator>) : BaseOperator(args) {

    override val argsDataTypes get() = listOf(listOf(DataType.Object))


    override val resultDataType get() = DataType.Class


    override fun compile(): CompilationResult {
        // Объявляем переменные
        val value = genVarName()
        val heads = ArrayList<String>()
        var completedRules = ""

        // Получаем аргументы
        val arg0 = arg(0)

        // Компилируем аргументы
        val compiledArg0 = arg0.compile()

        // Передаем завершенные правила дальше
        completedRules += compiledArg0.rules

        // Для всех результатов компиляции
        compiledArg0.bodies.forEach { head0 ->
            var head = head0
            head += genTriple(
                compiledArg0.value,
                genLink(JenaUtil.RDF_PREF, CLASS_PREDICATE_NAME),
                value
            ) + genNoValuePrim(
                value,
                genLink(JenaUtil.RDF_PREF, SUBCLASS_PREDICATE_NAME)
            )

            // Добавляем в массив
            heads.add(head)
        }

        return CompilationResult(value, heads, completedRules)
    }

    override fun clone(): Operator {
        val newArgs = ArrayList<Operator>()

        args.forEach { arg ->
            newArgs.add(arg.clone())
        }

        return GetClass(newArgs)
    }

    override fun clone(newArgs: List<Operator>): Operator {
        return GetClass(newArgs)
    }

    companion object {

        /**
         * Имя предиката, используемое при компиляции
         */
        private const val CLASS_PREDICATE_NAME = "type"

        /**
         * Имя предиката, используемое при компиляции
         */
        private const val SUBCLASS_PREDICATE_NAME = "subClassOf"
    }
}